doc/sphinx/README.rst

```diff
@@ -60,8 +60,11 @@ The signatures of most objects can be written using a succinct DSL for Coq notat
 ``{*, …}``, ``{+, …}``
   an optional or mandatory repeatable block, with repetitions separated by commas
 
-``%|``, ``%{``, …
-  an escaped character (rendered without the leading ``%``)
+``{| … | … | … }``
+  an alternative, indicating than one of multiple constructs can be used
+
+``%{``, ``%}``, ``%|``
+  an escaped character (rendered without the leading ``%``).  In most cases, escaping is not necessary.  In particular, the following expressions are all parsed as plain text, and do not need escaping: ``{ xyz }``, ``x |- y``.  But the following escapes *are* needed: ``{| a b %| c | d }``, ``all: %{``.  (We use ``%`` instead of the usual ``\`` because you'd have to type ``\`` twice in your reStructuredText file.)
```

Comment:
@cpitclaudel Can we refine this description to make it more specific?  I was updating doc_grammar to insert the escape characters and I had to guess a little (being lazy, I didn't do any experiments yet.)  If you give me the info I'll put in a PR.

True or false?
- If you're not inside a `{| ... }` construct then you don't need to escape the `|` or the `}`
- Don't use the `%|` form for a separator, e.g. use `{+|`, not `{+%|`.
- How about if `{+|` is inside a `{|`?
- Aside from the separator case, it's always valid to use the `%` escape for `|`, `{` and `}`?
- Does tokenization matter? E.g. if the grammar has a literal string `||`, is it OK in that form or should it be `%|%|` or `%||`?  (You allude to this with `|-` but don't state the rule.)

The easiest for doc_grammar is to always escape these characters if and only if they are entire tokens.

Reply:
Being more precise would be good, indeed :)

- If you're not inside a `{| ... }` construct then you don't need to escape the `|` or the `}`

I think you might need to escape the `}`.

- Don't use the `%|` form for a separator, e.g. use `{+|`, not `{+%|`.

Correct.

- How about if `{+|` is inside a `{|`?

That should work fine (the `|` should be understood as a separator, not as an alternative delimiter)

- Aside from the separator case, it's always valid to use the `%` escape for `|`, `{` and `}`?

I think so.

- Does tokenization matter? E.g. if the grammar has a literal string `||`, is it OK in that form or should it be `%|%|` or `%||`?  (You allude to this with `|-` but don't state the rule.)

I think it should be %|%|, but that's only needed within a `{| … }`.

Reply:
`make sphinx` takes foreeeeever to run on my system (about 35 minutes) and I've done that several times in the last couple hours.

I found 2 bugs:

- `{| {+, foo } }` gives a parse error.  This one we may want to fix because the following grammar items are like commands and so we probably would use the prodn/cmd syntax:

```
Infix @ne_lstring := @constr {| ( {+, @syntax_modifier } ) | } {? : @IDENT } 
Notation @lstring := @constr {| ( {+, @syntax_modifier } ) | } {? : @IDENT } 
Reserved Infix @ne_lstring {| ( {+, @syntax_modifier } ) | } 
Reserved Notation @ne_lstring {| ( {+, @syntax_modifier } ) | } 
```

- `{+|| xx }` gives a parse error, while other multicharacter separators such as ``{+of xx }`` work.  So far we don't use this in the grammar, so no fix needed yet.

Reply:
And these test cases worked:

```
.. prodn:: name1 ::= | || |-
.. prodn:: name1 += % %% %} %{ %| %|-
.. prodn:: name1 += {| x | y %| z }
.. prodn:: name1 += %{| %{? %{+,
.. prodn:: name1 += {+| xx }
.. prodn:: name1 += {+sep xx }
.. prodn:: name1 += { xyz }
```

![image](https://user-images.githubusercontent.com/1253341/59983062-d1139080-95cf-11e9-99c6-141ad9382c53.png)

Reply:
Are you sure the first one is a bug? Alternatives are expected to have at least two options, so I'm not too surprised that it fails; additionally, options are supposed to be non-empty (if empty options are permitted, you should use `{? …}` instead).

The second one is interesting.  I don't know if we currently have multi-character separators; I hadn't considered that case.

Reply:
For the first one, `{| {+, foo } | xxx }` parses.  So that shouldn't be an issue.  (I'll double check the rendering when the job completes.)

For the second one, I see a dozen multicharacter separators in the grammar, for some of which we may well want to use prodn/cmd.  "with" doesn't fit in the circle but that seems pretty minor:

```
gallina:
 {? @cumulativity_token } @private_token @finite_token {+with @inductive_definition } 
 Fixpoint {+with @rec_definition } 
 Let Fixpoint {+with @rec_definition } 
 CoFixpoint {+with @corec_definition } 
 Let CoFixpoint {+with @corec_definition } 
 Scheme {+with @scheme } 

command:
 Ltac {+with @ltac_tacdef_body } 
 Function {+with @function_rec_definition_loc } 
 Functional Scheme {+with @fun_scheme_arg } 
```